pkg/apiserver/middlewares/v1/api_key.go

```diff
@@ -46,22 +51,90 @@ func HashSHA512(str string) string {
 	return hashStr
 }
 
+func (a *APIKey) SetAllowedOUs(allowedOu []string) error {
+	for _, ou := range allowedOu {
+		//just drop empty strings, I guess ?
+		if ou == "" {
+			log.Warningf("Ignoring empty OU string in Bouncers Allowed OU (api-key)")
+			continue
+		}
+		a.AllowedOu = append(a.AllowedOu, ou)
+	}
+	log.Infof("%p Allowed Bouncers OU : %v", a, a.AllowedOu)
+	return nil
+}
+
 func (a *APIKey) MiddlewareFunc() gin.HandlerFunc {
 	return func(c *gin.Context) {
-		val, ok := c.Request.Header[APIKeyHeader]
-		if !ok {
-			c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
-			c.Abort()
-			return
-		}
+		var bouncer *ent.Bouncer
+		var err error
 
-		hashStr := HashSHA512(val[0])
-		bouncer, err := a.DbClient.SelectBouncer(hashStr)
-		if err != nil {
-			log.Errorf("auth api key error: %s", err)
-			c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
-			c.Abort()
-			return
+		if c.Request.TLS != nil && len(c.Request.TLS.PeerCertificates) > 0 {
+			validCert, extractedCN, err := ValidateCert(c, a.AllowedOu)
+			if !validCert {
+				c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+				c.Abort()
+				return
+			}
+			if err != nil {
+				log.Error(err)
+				c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+				c.Abort()
+				return
+			}
+			bouncerName := fmt.Sprintf("%s@%s", extractedCN, c.ClientIP())
+			log.Debugf("Trying to find bouncer %s for cert", bouncerName)
+			bouncer, err = a.DbClient.SelectBouncerByName(bouncerName)
+			if ent.IsNotFound(err) {
+				//Because we have a valid cert, automatically create the bouncer in the database if it does not exist
+				//Set a random API key, but it will never be used
+				//To fix
+				apiKey, err := GenerateAPIKey(64)
```

Comment:
we might not even want to create an api key ?

Reply:
This doesn't bother me that much...